python/ray/experimental/channel/nccl_group.py

```diff
@@ -0,0 +1,84 @@
+from typing import Dict, Optional
+
+import ray
+from ray.util.collective.collective_group import nccl_util
+
+try:
+    import torch
+except ImportError:
+    torch = None
+
+
+class _NcclGroup:
+    def __init__(
+        self,
+        world_size: int,
+        comm_id: int,
+        rank: Optional[int],
+        actor_ids_to_ranks: Dict[ray.ActorID, int],
+        cuda_stream: Optional[int],
+    ):
+        self._rank: Optional[int] = rank
+        if rank is not None:
+            assert ray.get_gpu_ids(), "NCCL actor has no GPUs assigned"
+
+            from cupy.cuda import nccl
+
+            self._comm = nccl.NcclCommunicator(world_size, comm_id, rank)
+        else:
+            # Driver does not have a rank.
+            self._comm = None
+        self._actor_ids_to_ranks = actor_ids_to_ranks
+
+        self._cuda_stream = cuda_stream
+
+        self._closed = False
+
+    def get_rank(self, actor: ray.actor.ActorHandle) -> int:
+        if actor._ray_actor_id not in self._actor_ids_to_ranks:
+            raise ValueError("Actor is not in the NCCL group.")
+        return self._actor_ids_to_ranks[actor._ray_actor_id]
+
+    def get_self_rank(self) -> int:
+        return self._rank
+
+    def send(self, value: torch.Tensor, peer_rank: int):
+        if self._closed:
+            raise IOError("NCCL group has been destroyed.")
+        # TODO(swang): Handle send/recv async NCCL errors such as network
+        # failures.
+        self._comm.send(
+            nccl_util.get_tensor_ptr(value),
+            value.numel(),
+            nccl_util.get_nccl_tensor_dtype(value),
+            peer_rank,
+            self._cuda_stream,
+        )
```

Comment:
is this blocking? If tensor is modified after this call, would that be okay?

Reply:
It is not blocking. I'll add some docs to send/recv on what is allowed. Basically if it's on the same stream (torch.cuda.current_stream()), it's okay.